uniffi/src/templates/ObjectTemplate.rs

```diff
@@ -50,14 +50,14 @@ lazy_static::lazy_static! {
         let mut err: ffi_support::ExternError = Default::default(); // XXX TODO: error handling!
         // If the method does not have the same signature as declared in the IDL, then
         // this attempt to call it will fail with a (somewhat) helpful compiler error.
-        UNIFFI_HANDLE_MAP_{{ obj.name()|upper }}.call_with_output_mut(&mut err, {{ meth.first_argument().name() }}, |obj| {
+        UNIFFI_HANDLE_MAP_{{ obj.name()|upper }}.{% match meth.error() %}{% when Some with (e) %}call_with_result_mut{% else %}call_with_output_mut{% endmatch %}(&mut err, {{ meth.first_argument().name() }}, |obj| {% match meth.error() %}{% when Some with (e) %}-> Result<{% match meth.ffi_func().return_type() %}{% when Some with (return_type) %}{{ return_type|ret_type_c }}{% else %}(){% endmatch %}>{% else %}{% endmatch %}{
             let _retval = {{ obj.name() }}::{{ meth.name() }}(
                 obj,
                 {%- for arg in meth.arguments() %}
                 {{ arg.name()|lift_rs(arg.type_()) }},
```

Comment:
Since we're here: note that `lift_rs` comes out as `<TYPE as ViaFfi>::try_from_ffi_value(VALUE).unwrap()`. If we're adding error handling, it would be nice to change that `unwrap` into a `?` or similar and have the `ViaFfi` conversion errors reported up to calling code in a sensible way.

Such errors should of course *never* happen in theory, because we control both sides of the FFI, so why would there ever be a conversion error? But you know what they say about theory and practice.

A strawperson suggestion: consider reserving a special error code (maybe `MAX_INT`?) for reporting these `ViaFfi` conversion errors, so that every function can implicitly throw a "something has gone horribly wrong in the bindings layer" error if something happens to go horribly wrong in the bindings layer.

Reply:
Alternatively, I suppose we could handle these conversion errors the same way we handle panics in the underlying rust code, and spin that work off into a separate "formalize and add tests for handling panics" issue.

Reply:
Hmmm yeah, this is an interesting problem, because it exists whether or not the function itself is expected to return a `Result`. I might spin off a different issue to deal with this... Here's what is going through my head right now:

- For functions that already return `Result`s we'd have to figure out a way to map the conversion error to the expected error or create a layer of indirection, and have both the user-defined error and the conversion errors map to that, then use that as the return value of the closure in `call_with_result_mut`
- For functions that don't return `Result`s, If we take the indirection route, then we just define the 'indirection' error anyways and change those calls to be `call_with_result_mut` even though the core rust function itself isn't expected to return a `Result`. 

I'm leaning towards having this as a follow-up bug, but want to make sure that I don't make decisions that make it tough to implement, thoughts?

Reply:
I also lean towards a separate issue. Let's continue to panic on conversion errors since they should never happen in practice, and rely on safe panic handling (which we need anyway!) to report them up to callers.  I filed https://github.com/rfk/uniffi-rs/issues/32 for followup.

 I think the main impacts on the current PR will be:

* All FFI functions would need to take an out `ExternError` for handling panics, which might influence how you do codegen for handling other errors.
* You need to avoid conflicting with reserved error codes (IIUC from [the docs](https://docs.rs/ffi-support/0.4.0/ffi_support/struct.ExternError.html) we reserve `0` for success and `-1` for reporting internal errors such as panics.

I can't think of any other corners you might back yourself in to as a result of this PR.